torch/csrc/jit/tensorexpr/bounds_overlap.cpp

```diff
@@ -6,6 +6,73 @@ namespace jit {
 namespace tensorexpr {
 namespace analysis {
 
+// Returns true if the given expression includes a "Sub" term.
+bool hasSub(ExprPtr e) {
+  return !NodeFinder<Sub>::find(e).empty();
+}
+
+// Returns true if the given expression includes a "Div" term.
+bool hasDiv(ExprPtr e) {
+  return !NodeFinder<Div>::find(e).empty();
+}
+
+// Returns true if all immediates in the given expression are positive.
+bool allImmsArePositive(ExprPtr e) {
+#define TYPE_CASE(Type, Name)                                 \
+  {                                                           \
+    auto imms = NodeFinder<Name##Imm>::find(e);               \
+    for (const auto& imm : imms) {                            \
+      if (immediateIsNegative(imm) || immediateIsZero(imm)) { \
+        return false;                                         \
+      }                                                       \
+    }                                                         \
+  }
+  AT_FORALL_SCALAR_TYPES_AND2(Half, BFloat16, TYPE_CASE);
+#undef TYPE_CASE
+  {
+    auto imms = NodeFinder<BoolImm>::find(e);
+    for (const auto& imm : imms) {
+      if (immediateIsZero(imm)) {
+        return false;
+      }
+    }
+  }
+  return true;
+}
+
+// Returns true if the given expression is guaranteed to be positive.
+//
+// If the given expression is constant, this checks if the constant value is
+// positive.
+//
+// If the given expression is not constant, we use the following conservative
+// approximation to identify those that are guaranteed to be positive.
+//
+// Assumption:
+//   * The variables in the given expression are always positive.
+//
+// The only way for the given expression to be negative is if it includes
+// at least one of:
+//     * "subtract" term.
```

Comment:
In theory bitwise ops plus cast can also result in a negative value. But I only mention it for the sake of completeness, I'm not sure we really need to add them here. Same with shifts for zero values.